### PR TITLE
fix: Multi-select dropdown not closing properly

### LINKED ISSUE #12458: [Bug]: Multiselect Dropdown doesn't close when the label is on the side
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

The Multiselect widget displays weirdly in compact mode, i.e when the label is on its side

### Steps To Reproduce

Just drag a Multiselect Widget 

Make it quite Long.
Then add a Label to be on the side of the widget

[LOOM DEMO](https://www.loom.com/share/dc5c31b3ac2b4c8c9526a5828e5a63f9)

### Public Sample App

https://release.app.appsmith.com/performance-dashboard-beta/tracking-by-metric-6245d30320d39e42471bf99d

### Version

Cloud

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

1. Switched from using a memorized value for DropdownWidth to making it a state. LabelRef was still null after render.
2. Also Increased the specificity of MultiSelect and MultiTreeSelect, Issues arise when two of these widgets are on the canvas

Fixes #12458

## Type of change

- Bugfix 

## How Has This Been Tested?

- Visual

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: fix/Multiselect-style-specificity 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 56.18 **(0)** | 37.66 **(-0.01)** | 35.71 **(0.01)** | 56.42 **(-0.01)**
 :red_circle: | app/client/src/utils/WorkerUtil.ts | 88.98 **(-0.78)** | 70.59 **(-1.96)** | 100 **(0)** | 92.38 **(-0.95)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :green_circle: | app/client/src/widgets/WidgetUtils.ts | 86.41 **(0.14)** | 61.9 **(0)** | 62.5 **(0)** | 83.53 **(0.2)**
 :green_circle: | app/client/src/widgets/MultiSelectTreeWidget/component/index.styled.tsx | 47.62 **(2.17)** | 25 **(0)** | 0 **(0)** | 51.43 **(1.43)**
 :red_circle: | app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx | 15.85 **(-0.82)** | 0 **(0)** | 0 **(0)** | 17.81 **(-0.5)**
 :green_circle: | app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx | 50 **(2.5)** | 31.25 **(0)** | 0 **(0)** | 54.55 **(1.61)**
 :red_circle: | app/client/src/widgets/MultiSelectWidgetV2/component/index.tsx | 11.4 **(-0.42)** | 0 **(0)** | 0 **(0)** | 13 **(-0.27)**
 :green_circle: | app/client/src/widgets/SelectWidget/component/index.tsx | 15.32 **(0.8)** | 0 **(0)** | 4.55 **(0)** | 15.93 **(0.89)**
 :red_circle: | app/client/src/widgets/SelectWidget/widget/index.tsx | 54.93 **(0.14)** | 22.22 **(-0.99)** | 50 **(0)** | 55.88 **(0.17)**
 :red_circle: | app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx | 16.05 **(-0.83)** | 0 **(0)** | 0 **(0)** | 18.06 **(-0.51)**</details>

### CODE REVIEW

File: app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx [hunk 1]
```diff
@@ -157,7 +157,38 @@ export const DropdownStyles = createGlobalStyle<{
 }>`
 ${({ dropDownWidth, id }) => `
   .multiselect-popover-width-${id} {
-    min-width: ${dropDownWidth}px !important;
+    width: ${dropDownWidth}px !important;
+
+    &.rc-select-dropdown-hidden {
+      display: none !important;
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx:163]: Please refrain from using !important

File: app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx [hunk 2]
```diff
@@ -157,7 +157,38 @@ export const DropdownStyles = createGlobalStyle<{
...
+      background: #fff;
+      width: auto;
+      border-radius: 0px;
+      margin-top: 5px;
+      background: white;
+      box-shadow: 0 6px 20px 0px rgba(0, 0, 0, 0.15) !important;
+      overflow-x: scroll;
+      > div {
+          min-width: ${dropDownWidth}px;
+        }
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx:177]: alignment is off

File: app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx [hunk 3]
```diff
@@ -283,35 +314,10 @@ ${({ dropDownWidth, id }) => `
...
-    width: 100%;
-    height: 38px;
-  }
-  .rc-select-item-option-state {
-    .bp3-control.bp3-checkbox {
-      margin-bottom: 0;
-    }
-  }
 }
+
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx:320]: remove

**Review by sbalaji1192 — COMMENTED** [review]
Please add a cypress case as well if these changes are testable.

File: app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx (hunk 1 above)
Comment (Tooluloope) [reply app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx]: It's kinda necessary to override the default RC-select config

File: app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx [hunk 4]
```diff
@@ -151,15 +152,20 @@ function MultiTreeSelectComponent({
...
+      );
+      return;
     }
-    const parentWidth = width - WidgetContainerDiff;
-    return parentWidth > dropDownWidth ? parentWidth : dropDownWidth;
-  }, [compactMode, dropDownWidth, width, labelRef.current]);
+    setMemoDropDownWidth(
+      parentWidth > dropDownWidth ? parentWidth : dropDownWidth,
+    );
+  }, [compactMode, dropDownWidth, width, labelText]);
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx:168]: we dont need to add `labelRef` ?

File: app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx [hunk 5]
```diff
@@ -157,6 +157,7 @@ export const DropdownStyles = createGlobalStyle<{
 }>`
 ${({ dropDownWidth, id }) => `
   .multiselect-popover-width-${id} {
+    width: ${dropDownWidth}px !important;
```
Comment (sbalaji1192) [inline app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx:160]: use && if we can

File: app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx [hunk 6]
```diff
@@ -152,15 +153,20 @@ function SingleSelectTreeComponent({
...
+      );
+      return;
     }
-    const parentWidth = width - WidgetContainerDiff;
-    return parentWidth > dropDownWidth ? parentWidth : dropDownWidth;
-  }, [compactMode, dropDownWidth, width, labelRef.current]);
+    setMemoDropDownWidth(
+      parentWidth > dropDownWidth ? parentWidth : dropDownWidth,
+    );
+  }, [compactMode, dropDownWidth, width, labelText]);
```
[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx:169]: No, I already have `labelText` in the dependency

File: app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx (hunk 4 above)
Comment (Tooluloope) [reply app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx:168]: No, I already have `labelText` in the dependency

File: app/client/src/widgets/MultiSelectTreeWidget/component/index.styled.tsx [hunk 7]
```diff
@@ -84,6 +84,7 @@ export const DropdownStyles = createGlobalStyle<{
 ${({ dropDownWidth, id }) => `
   .multiselecttree-popover-width-${id} {
     min-width: ${dropDownWidth}px !important;
+    width: ${dropDownWidth}px !important;
```
[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/MultiSelectTreeWidget/component/index.styled.tsx:87]: The issue here is that RC-select uses inline style for width and min-width, using more specificity would require using important.

Screenshot 2022-04-05 at 07 25 36

**Review by sbalaji1192 — APPROVED** [review]

**Comment by vivekverma2312** [comment]
Tested the issue and work as expected.

**Pull request merged**